Speed up batou age interaction

PhilTaken opened the pull request:
The most relevant commit is 8d569b295d8afd9c5b51aeb37f5b93534c85b465.

Due to the previous implementation working with generous amounts of temporary files (especially so when using `age-diffable`), the IO overhead from reading from and writing to these files is large enough to severely slow down secret handling operations.

This PR implements in-memory handling of deployment secrets via bindings to the [rage](https://github.com/str4d/rage) library, a rust implementation of the age spec. This leads to massively improved secret handling performance.

While this PR does introduce some added requirements on external libraries, it also relaxes the dependence on the (r)age cli.

When implementing this PR, I discovered a few bugs in the test implementation such as improper isolation of calls to `git` and `gpg` having issues creating an agent socket in nested folder structures as well as a dependence on a very specific terminal width where a bunch of tests would fail due to very strict string comparison. I have added these changes to this PR.

N.B.: due to limitations around handling of RSA keys in the cryptography library that rage depends on, RSA keys with moduli > 4096 are (currently) not supported via the bindings. In these cases, the implementation attempts to fall back to calling the `age` binary - see commit 000b98af551530d36f99a3ef1263c36858248893. Since `rage` is incompatible with these RSA keys anyways, the previous command discovery implementation could be simplified.

PhilTaken committed:
improve secret handling performance

Massively improve secret handling performance by avoiding large
amounts of writing to / reading from temporary files.

This solves a lot of complexity around the age call because it
does not require piping the ssh key's password into a subprocess' stdin
conditionally.

Instead, this can be handled using `cryptography` to handle the
password-protected ssh key and `pyrage`, python bindings to the rage
library which supports de/encrypting in-memory.

Another benefit of this implementation is a relaxed requirement (now
optional) on the `age` CLI since the bindings to the rage library
provide a full implementation out of the box.

PhilTaken committed:
update example lockfiles

PhilTaken committed:
fix tests: remove assert on fixed terminal width

PhilTaken committed:
fix tests: isolate git more from the base system

The test suite provides some git configuration, additional
custom git configuration should not cause test cases to fail.

PhilTaken committed:
add fallback behaviour for age encryption

recipient errors can occur when using rsa keys with module > 4096 as
recipients with rage, age can handle those

PhilTaken committed:
add changelog

PhilTaken committed:
fix type hinting: python < 3.9 cannot subscript `type`

elikoga commented:
Calling out to the rage library is probably the right call for performance here. I distinctly remember being very frustrated during development of the age binary integration.

PhilTaken commented:
The largest performance hog now is the call out to `op` to read the password, at around 300-600ms each the impact is pretty significant if you want to use `batou secrets decrypttostdout` in git diffs. It currently sits around 1 second total for each side of the diff for each commit. 

Certainly usable but could be significantly improved imho.

PhilTaken committed:
breaking: drop python 3.7 compatibility

zagy commented on CHANGES.d/20250217_112817_ph_speedup_batou_age_interaction.md:
This needs to note that 3.7 is no longer supported.

zagy commented on src/batou/tests/test_deploy.py:
```diff
@@ -31,7 +31,7 @@ def test_main_with_errors(capsys):
         """\
 batou/2... (cpython 3...)
-================================== Preparing ===================================
+... Preparing ...
```
It's not quite clear why this changes when changing the encryption?

PhilTaken replied:
This is not directly related to the encryption but was required so I can run the test suite on my machine to test all other changes. The specific commit is d4caccaf4239dc1d7909859a2fb1f63353c62a3b. I can move these out to a separate PR now if you want.

zagy commented:
I also get an error in our standard deploy container, meaning there is another dependency which might not be available everywhere (think switches). While a remote system does not *need* do decrypt (since that happens locally) the dependency would be required. We need to re-think this, I suppose.

```
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'error'
  error: subprocess-exited-with-error

  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [6 lines of output]

      Cargo, the Rust package manager, is not installed or is not on PATH.
      This package requires Rust and Cargo to compile extensions. Install it through
      the system's package manager or via https://rustup.rs/

      Checking for Rust toolchain....
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

PhilTaken committed:
add note about dropping python 3.7

leona-ya commented on src/batou/secrets/encryption.py:
```diff
@@ -258,13 +263,43 @@ def get_identities():
+                key_content = f.read()
+            try:
+                priv_key = serialization.load_ssh_private_key(key_content, None)
+            except ValueError:
```
This actually should be 

```suggestion
            except (ValueError, TypeError):
```

I ran into an issue where `batou secrets edit staging` failed with the error message thrown inside `cryptography`
```
Key is password-protected, but password was not provided.
```
that is a `TypeError`

PhilTaken committed:
catch more errors during age encryption

PhilTaken commented:
I will attempt to loosen the dependency on pyrage by adding a bit of yes/no logic in the encryption code such that systems that either don't need pyrage at all or don't have a pre-built pyrage wheel available are not deprecated with this PR

PhilTaken committed:
provide a transparent switch back to the legacy version

if pyrage is not installed
also adds tests for both cases and sets up tox in such a way that the
"default" py envs do not have the optional dependency installed while a
second one does. Both are tested in CI for maximum coverage

PhilTaken commented:
rebased onto current main

PhilTaken committed:
wip: add mercurial for testing

PhilTaken committed:
tests: pass expected main branch to git init

git init falls back to master in some versions
also: explicit > implicit

PhilTaken committed:
remove unmaintained init-template

batou init was removed in d7fde5e5828eb40502dfd5621cc1b3ff68aa3c02

PhilTaken committed:
fix some more tests

by adding subversion + passing the assumed default branch to git init

PhilTaken committed:
fix more tests

PhilTaken committed:
add pyrage dependency group and tox env

PhilTaken committed:
ci: limit push runs to the main branch

PhilTaken committed:
ci: limit bootstrap and nix test runs to pushes to main and prs

PhilTaken committed:
ci: limit pre-commit push runs to main

PhilTaken committed:
ci: remove setup-python matrix in unit test since uv handles python versions

PhilTaken committed:
fix: remove leaky python passthru from pkgs.mercurial

PhilTaken committed:
tests: fix python3.14 compatibility

PhilTaken committed:
examples: migrate to pyproject

PhilTaken committed:
ci: fix the bootstrap script

PhilTaken committed:
tests: add uv.lock in examples

PhilTaken committed:
ci: fix flake check
